atc/builds/tracker.go

```diff
@@ -6,14 +6,27 @@ import (
 
 	"code.cloudfoundry.org/lager/lagerctx"
 	"github.com/concourse/concourse/atc/db"
-	"github.com/concourse/concourse/atc/engine"
 	"github.com/concourse/concourse/atc/metric"
 	"github.com/concourse/concourse/atc/util"
 )
 
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+//go:generate counterfeiter . Engine
+type Engine interface {
```

Comment:
Looks like redefine `Engine` here just for defining a no-return `Runnable`. I feel not worthy doing so. If we want to silently ignore error returned from a run, `engine` may just not checking return value of `Run`. Adding a duplicate `Engine` and `Runnable` adds some confusion.

Reply:
At one point, I think it may have been an import cycle - but even with no import cycle, the [Go authors recommend](https://github.com/golang/go/wiki/CodeReviewComments#interfaces) defining interfaces where they're used, rather than where they're implemented:

> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.

As you've no doubt noticed, we do a bad job following that advice, and honestly it's caused some pain 😁  but this tends to be a more sane way of defining dependencies.

Reply:
I actually would argue that it makes more sense to have it in the build tracker, this is because I found that the best practice is to create the interfaces where it is used rather where it is implemented. This makes it so that we can have more small and condensed interfaces while the implementation of the objects that conform to the interface be less tied to that interface. The interface should be directly tied to only what that specific component needs out of the object.

Throughout our codebase, we have interfaces really closely tied to the objects that implement them (really good example is the db package). But the only reason why we do that is to be able to test easily using fakes. I think if we can avoid tying the interface with the object, that would be best. What do you think about this?

Reply:
Yeah, make sense.